ev3dev2/motor.py

```diff
@@ -1924,6 +2037,181 @@ def get_speed_steering(self, steering, speed):
         return (left_speed, right_speed)
 
 
+class MoveDifferential(MoveTank):
+    """
+    For all methods speed can be an integer or a SpeedInteger object
+    """
+
+    def __init__(self, left_motor_port, right_motor_port, wheel_class, wheel_distance_mm, desc=None, motor_class=LargeMotor):
+        MoveTank.__init__(self, left_motor_port, right_motor_port, desc, motor_class)
+        self.wheel = wheel_class()
+
+        # The distance between the mid point of the wheels.  You may need to
+        # do some test drives to find the correct value for your robot.  It
+        # is not as simple as measuring the distance between the midpoints of
+        # the two wheels. The weight of the robot, center of gravity, etc come
+        # into play.  You can use utils/move_differential.py to call on_arc_left()
+        # to do some test drives in 1-foot circles.  Tweak wheel_distam_mm until
+        # your robot can drive in a 1-foot circle and stop exactly where it started.
+        self.wheel_distance_mm = wheel_distance_mm
+
+        # The circumference of the circle made if this robot were to rotate in place
+        self.circumference_mm = self.wheel_distance_mm * pi
+
+        self.min_circle_radius_mm = self.wheel_distance_mm / 2
+
+    def on_for_distance(self, speed, distance_mm, brake=True, block=True):
+        """
+        Drive forward distance_mm
+        """
+        abort = False
+
+        if not speed:
+            log.warning("{}: speed is invalid ({}), tank will not move".format(self, speed))
+            abort = True
+
+        if not distance_mm:
+            log.warning("{}: distance_mm is invalid ({}), tank will not move".format(self, distance_mm))
+            abort = True
+
+        if abort:
+            self._set_brake(brake)
+            return
+
+        rotations = distance_mm / self.wheel.circumference_mm
+        log.debug("%s: on_for_rotations distance_mm %s, rotations %s, speed %s" % (self, distance_mm, rotations, speed))
+
+        MoveTank.on_for_rotations(self, speed, speed, rotations, brake, block)
+
+    def _on_arc(self, speed, radius_mm, distance_mm, brake, block, arc_right):
+        """
+        Drive in a circle with 'radius' for 'distance'
+        """
+        abort = False
+
+        if not speed:
+            log.warning("{}: speed is invalid ({}), tank will not move".format(self, speed))
+            abort = True
```

Comment:
Do we really need the additional machinery for checking validity then aborting? I think it's fine to get one error then to fix that and get another if needed.

Reply:
Honestly I would be fine with just a bunch of asserts here but that means we don’t call set_brake when something goes wrong. I could live with that though. Thoughts?

Reply:
There are other places where we only have asserts and others where we log a warning and call set_brake.  We should make these consistent. My vote would be for asserts everywhere.

Reply:
You could create a utility method which does an assertion and stops the motors if the assert fails.

I like the behavior where it stops the motor if an error occurs, but maybe it isn't necessary and we can go for the simple option instead.

Reply:
There are probably a ton of places where we could raise an exception without first stopping the motors :( Let’s just do asserts and keep it nice and simple. The user will pretty always need their program to catch exceptions and stop the motors as a result.